zebra/dplane_fpm_nl.c

```diff
@@ -491,14 +489,79 @@ static void fpm_read(struct thread *t)
 		FPM_RECONNECT(fnc);
 		return;
 	}
-	stream_reset(fnc->ibuf);
+
+	/* Schedule the next read */
+	thread_add_read(fnc->fthread->master, fpm_read, fnc, fnc->socket,
+			&fnc->t_read);
+	/* We've got an interruption. */
+	if (rv == -2) {
+		/* Schedule next read. */
+		thread_add_read(fnc->fthread->master, fpm_read, fnc,
+				fnc->socket, &fnc->t_read);
+		return;
+	}
 
 	/* Account all bytes read. */
 	atomic_fetch_add_explicit(&fnc->counters.bytes_read, rv,
 				  memory_order_relaxed);
 
-	thread_add_read(fnc->fthread->master, fpm_read, fnc, fnc->socket,
-			&fnc->t_read);
+	while (rv) {
+		if (rv < (ssize_t)FPM_MSG_HDR_LEN) {
+			stream_reset(fnc->ibuf);
+			zlog_warn(
+				"%s: Received %zd bytes but cannot even read fpm header, ignoring",
+				__func__, rv);
+			return;
+		}
+
+		fpm.version = stream_getc(fnc->ibuf);
+		fpm.msg_type = stream_getc(fnc->ibuf);
+		fpm.msg_len = stream_getw(fnc->ibuf);
+
+		if (fpm.version != FPM_PROTO_VERSION &&
+		    fpm.msg_type != FPM_MSG_TYPE_NETLINK) {
+			stream_reset(fnc->ibuf);
+			zlog_warn(
+				"%s: Received version/msg_type %u/%u, expected 1/1",
+				__func__, fpm.version, fpm.msg_type);
+			return;
+		}
+
+		if (fpm.msg_len != rv) {
+			stream_reset(fnc->ibuf);
+			zlog_warn(
+				"%s: Received message length %u but read only got %zd, ignoring message",
+				__func__, fpm.msg_len, rv);
+		}
+
+		/*
+		 * Place the data from the stream into a buffer
+		 */
+		hdr = (struct nlmsghdr *)buf;
+		stream_get(buf, fnc->ibuf, fpm.msg_len - FPM_MSG_HDR_LEN);
+		rv -= fpm.msg_len;
+
+		switch (hdr->nlmsg_type) {
+		case RTM_NEWROUTE:
+			ctx = dplane_ctx_alloc();
+			dplane_ctx_set_op(ctx, DPLANE_OP_ROUTE_NOTIFY);
+			if (netlink_route_change_read_unicast_internal(
+				    hdr, 0, false, ctx) != 1) {
+				dplane_ctx_fini(&ctx);
+				stream_reset(fnc->ibuf);
```

Comment:
`stream_get` consumes the data, but the `getp` will still be in the middle of the stream. I recommend calling `stream_pulldown` in `ibuf` instead of `stream_reset`.

Reply:
fixed